source/common/upstream/load_stats_reporter.cc

```diff
@@ -0,0 +1,139 @@
+    handleFailure();
+    return;
+  }
+
+  request_.mutable_cluster_stats()->Clear();
+  sendLoadStatsRequest();
+}
+
+void LoadStatsReporter::sendLoadStatsRequest() {
+  request_.mutable_cluster_stats()->Clear();
+  for (const std::string& cluster_name : clusters_) {
+    auto cluster_info_map = cm_.clusters();
+    auto it = cluster_info_map.find(cluster_name);
+    if (it == cluster_info_map.end()) {
+      ENVOY_LOG(debug, "Cluster {} does not exist", cluster_name);
+      continue;
+    }
+    auto& cluster = it->second.get();
+    auto* cluster_stats = request_.add_cluster_stats();
+    cluster_stats->set_cluster_name(cluster_name);
+    for (auto& hosts : cluster.hostsPerLocality()) {
+      auto* locality_stats = cluster_stats->add_upstream_locality_stats();
+      ASSERT(hosts.size() > 0);
+      locality_stats->mutable_locality()->MergeFrom(hosts[0]->locality());
+      uint64_t rq_success = 0;
+      uint64_t rq_error = 0;
+      uint64_t rq_active = 0;
+      for (auto host : hosts) {
+        rq_success += host->stats().rq_success_.latch();
+        rq_error += host->stats().rq_error_.latch();
+        rq_active += host->stats().rq_active_.value();
+      }
+      locality_stats->set_total_successful_requests(rq_success);
+      locality_stats->set_total_error_requests(rq_error);
+      locality_stats->set_total_requests_in_progress(rq_active);
+    }
+    cluster_stats->set_total_dropped_requests(
+        cluster.info()->loadReportStats().upstream_rq_dropped_.latch());
+  }
+
+  ENVOY_LOG(trace, "Sending LoadStatsRequest: {}", request_.DebugString());
+  stream_->sendMessage(request_, false);
+  stats_.responses_.inc();
+}
+
+void LoadStatsReporter::handleFailure() {
+  ENVOY_LOG(warn, "Load reporter stats stream/connection failure, will retry in {} ms.",
+            RETRY_DELAY_MS);
+  stats_.errors_.inc();
+  setRetryTimer();
+}
+
+void LoadStatsReporter::onCreateInitialMetadata(Http::HeaderMap& metadata) {
+  UNREFERENCED_PARAMETER(metadata);
+}
+
+void LoadStatsReporter::onReceiveInitialMetadata(Http::HeaderMapPtr&& metadata) {
+  UNREFERENCED_PARAMETER(metadata);
+}
+
+void LoadStatsReporter::onReceiveMessage(
+    std::unique_ptr<envoy::api::v2::LoadStatsResponse>&& message) {
+  ENVOY_LOG(debug, "New load report epoch: {}", message->DebugString());
+  clusters_.clear();
+  // Reset stats for all hosts in clusters we are tracking.
+  for (const std::string& cluster_name : message->clusters()) {
+    clusters_.emplace_back(cluster_name);
+    auto cluster_info_map = cm_.clusters();
+    auto it = cluster_info_map.find(cluster_name);
+    if (it == cluster_info_map.end()) {
+      continue;
+    }
+    auto& cluster = it->second.get();
+    for (auto host : cluster.hosts()) {
+      host->stats().rq_success_.latch();
+      host->stats().rq_error_.latch();
+    }
+    cluster.info()->loadReportStats().upstream_rq_dropped_.latch();
+  }
+  stats_.requests_.inc();
+  response_timer_->enableTimer(std::chrono::milliseconds(
```

Comment:
Is it worth warning or doing anything if the server sends a 2nd message to reconfigure before we sent load reporting data? Is that normal? Do we have a test case for that? (I think code works fine in that case fwiw).

Reply:
Will add a test.